### PR TITLE
Split sections support

### LINKED ISSUE #4819: Split sections is not supported
Cabal has a `--enable-split-objs` flag yet no `--enable-split-sections`. Note that GHC recommends using section splitting in favor of object splitting wherever possible.

### PR DESCRIPTION
**Pull request by bgamari**
* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Add --enable-split-sections flag and pipe it through to the GHC backend. Note
that some of the implementation here could be made a bit more precise:
-split-sections and -split-objs are mutually exlusive yet the types don't
currently reflect this.

Fixes #4819.

### CODE REVIEW

File: Cabal/Distribution/Simple/GHC/Internal.hs [hunk 1]
```diff
@@ -319,6 +319,7 @@ componentGhcOptions verbosity implInfo lbi bi clbi odir =
       ghcOptWarnMissingHomeModules = toFlag $ flagWarnMissingHomeModules implInfo,
       ghcOptPackageDBs      = withPackageDB lbi,
       ghcOptPackages        = toNubListR $ mkGhcOptPackages clbi,
+      ghcOptSplitSections   = toFlag (splitSections lbi),
       ghcOptSplitObjs       = toFlag (splitObjs lbi),
```
Comment (bgamari) [inline Cabal/Distribution/Simple/GHC/Internal.hs]: One potential refactoring would be to fold these two fields into one `Maybe CodeShrinkScheme` with,
```haskell
data CodeShrinkScheme = SplitObjs | SplitSections
```
I'm not sure this is worth the trouble though.

File: Cabal/Distribution/Simple/Configure.hs [hunk 2]
```diff
@@ -606,11 +606,31 @@ configure (pkg_descr0, pbi) cfg = do
             installedPackageSet
             comp
 
+    -- Decide if we're going to compile with split sections.
+    split_sections :: Bool <-
+       if not (fromFlag $ configSplitSections cfg)
+            then return False
+            else case compilerFlavor comp of
+                        GHC | compilerVersion comp >= mkVersion [8,0]
```
Comment (duog) [inline Cabal/Distribution/Simple/Configure.hs]: I think this should be 8.2?

Comment (bgamari) [reply Cabal/Distribution/Simple/Configure.hs]: 8.0 also supports `-split-sections`. We merely started building binary distributions with ``split-sections` in 8.2

**Comment by 23Skidoo** [comment]
Travis is red, failures look legit.

File: Cabal/Distribution/Simple/GHC/Internal.hs (hunk 1 above)
Comment (23Skidoo) [reply Cabal/Distribution/Simple/GHC/Internal.hs]: Yeah, don't bother. IMO `-split-objs` should be removed at some point.

**Comment by 23Skidoo** [comment]
Now there's a merge conflict.

**Comment by bgamari** [comment]
How does this look?

**Comment by angerman** [comment]
appveyor failed with: 

> unrecognized 'configure' option `--disable-split-sections'

which I believe originates from `commandParseArgs` in `Distribution.Simple.Command`.
Not sure why though.

**Comment by 23Skidoo** [comment]
Maybe `filterConfigureFlags` wasn't updated?

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Comment by phadej** [comment]
Hmm. Seems that `build-type: Custom` `./Setup` doesn't know `--disable-split-sections` / `--enable-split-sections`.

@bgamari I assume you didn't try that scenario?

**Comment by phadej** [comment]
I can reproduce the failure with custom setup by eg. changing `primitive` to be `build-type: Custom`:

```
[FL973] ~/mess/primitive-0.6.2.0 master % git diff               
diff --git a/primitive.cabal b/primitive.cabal
index 138183e..8cd30aa 100644
--- a/primitive.cabal
+++ b/primitive.cabal
@@ -12,7 +12,7 @@ Bug-Reports:    https://github.com/haskell/primitive/issues
 Category:       Data
 Synopsis:       Primitive memory-related operations
 Cabal-Version:  >= 1.10
-Build-Type:     Simple
+Build-Type:     Custom^M
 Description:    This package provides various primitive memory-related operations.
 
 Extra-Source-Files: changelog.md
@@ -24,6 +24,9 @@ Tested-With:
   GHC == 7.10.3,
   GHC == 8.0.1
 
+custom-setup^M
+  setup-depends: base, Cabal^M
+^M
 Library
   Default-Language: Haskell2010
   Other-Extensions:
[FL973] ~/mess/primitive-0.6.2.0 master % ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.4.0.20171214
[FL973] ~/mess/primitive-0.6.2.0 master % cabal new-build --allow-newer='*:base'
Resolving dependencies...
Build profile: -w ghc-8.4.0.20171214 -O1
In order, the following will be built (use -v for more details):
 - primitive-0.6.2.0 (lib:primitive) (first run)
[1 of 1] Compiling Main             ( /home/ogre/mess/primitive-0.6.2.0/dist-newstyle/build/x86_64-linux/ghc-8.4.0.20171214/primitive-0.6.2.0/setup/setup.hs, /home/ogre/mess/primitive-0.6.2.0/dist-newstyle/build/x86_64-linux/ghc-8.4.0.20171214/primitive-0.6.2.0/setup/Main.o )
Linking /home/ogre/mess/primitive-0.6.2.0/dist-newstyle/build/x86_64-linux/ghc-8.4.0.20171214/primitive-0.6.2.0/setup/setup ...
unrecognized 'configure' option `--disable-split-sections'
```

**Comment by phadej** [comment]
status: maybe the issue is because my `cabal-install` is newer then `Cabal` bundled with `8.4.0.20171214`. Let's see.

**Comment by 23Skidoo** [comment]
`libraries/Cabal` submodule on the 8.4 branch is set to 97c66f2c7698f0aea4277acb66b918b7341b3d01, which is before this patch was merged.

/cc @bgamari

**Pull request merged**